README
NAME
    Class::Accessor::Grouped - Lets you build groups of accessors

SYNOPSIS
DESCRIPTION
    This class lets you build groups of accessors that will call different
    getters and setters.

METHODS
  mk_group_accessors
    Arguments: $group, @fieldspec
        Returns: none

    Creates a set of accessors in a given group.

    $group is the name of the accessor group for the generated accessors;
    they will call get_$group($field) on get and set_$group($field, $value)
    on set.

    If you want to mimic Class::Accessor's mk_accessors $group has to be
    'simple' to tell Class::Accessor::Grouped to use its own get_simple and
    set_simple methods.

    @fieldspec is a list of field/accessor names; if a fieldspec is a scalar
    this is used as both field and accessor name, if a listref it is
    expected to be of the form [ $accessor, $field ].

  mk_group_ro_accessors
    Arguments: $group, @fieldspec
        Returns: none

    Creates a set of read only accessors in a given group. Identical to
    <L:/mk_group_accessors> but accessors will throw an error if passed a
    value rather than setting the value.

  mk_group_wo_accessors
    Arguments: $group, @fieldspec
        Returns: none

    Creates a set of write only accessors in a given group. Identical to
    <L:/mk_group_accessors> but accessors will throw an error if not passed
    a value rather than getting the value.

  make_group_accessor
    Arguments: $group, $field
        Returns: $sub (\CODE)

    Returns a single accessor in a given group; called by mk_group_accessors
    for each entry in @fieldspec.

  make_group_ro_accessor
    Arguments: $group, $field
        Returns: $sub (\CODE)

    Returns a single read-only accessor in a given group; called by
    mk_group_ro_accessors for each entry in @fieldspec.

  make_group_wo_accessor
    Arguments: $group, $field
        Returns: $sub (\CODE)

    Returns a single write-only accessor in a given group; called by
    mk_group_wo_accessors for each entry in @fieldspec.

  get_simple
    Arguments: $field
        Returns: $value

    Simple getter for hash-based objects which returns the value for the
    field name passed as an argument.

  set_simple
    Arguments: $field, $new_value
        Returns: $new_value

    Simple setter for hash-based objects which sets and then returns the
    value for the field name passed as an argument.

  get_inherited
    Arguments: $field
        Returns: $value

    Simple getter for Classes and hash-based objects which returns the value
    for the field name passed as an argument. This behaves much like
    Class::Data::Accessor where the field can be set in a base class,
    inherited and changed in subclasses, and inherited and changed for
    object instances.

  set_inherited
    Arguments: $field, $new_value
        Returns: $new_value

    Simple setter for Classes and hash-based objects which sets and then
    returns the value for the field name passed as an argument. When called
    on a hash-based object it will set the appropriate hash key value. When
    called on a class, it will set a class level variable.

    Note:: This method will die if you try to set an object variable on a
    non hash-based object.

  get_component_class
    Arguments: $field
        Returns: $value

    Gets the value of the specified component class.

        __PACKAGE__->mk_group_accessors('component_class' => 'result_class');

        $self->result_class->method();

        ## same as
        $self->get_component_class('result_class')->method();

  set_component_class
    Arguments: $field, $class
        Returns: $new_value

    Inherited accessor that automatically loads the specified class before
    setting it. This method will die if the specified class could not be
    loaded.

        __PACKAGE__->mk_group_accessors('component_class' => 'result_class');
        __PACKAGE__->result_class('MyClass');

        $self->result_class->method();

  get_super_paths
    Returns a list of 'parent' or 'super' class names that the current class
    inherited from.

AUTHORS
    Matt S. Trout <mst@shadowcatsystems.co.uk> Christopher H. Laco
    <claco@chrislaco.com>

    With contributions from:

    Guillermo Roditi <groditi@cpan.org>

LICENSE
    You may distribute this code under the same terms as Perl itself.



Changes
Revision history for Class::Accessor::Grouped.

    - releasing 0.10013_01 as stable

0.10013_01 2018-05-28 13:37 (UTC)
    - convert packaging from Module::Install to plain ExtUtils::MakeMaker
    - fix running Makefile.PL when @INC does not contain '.' (perl 5.26).
    - don't run author tests at all for user installs

0.10012 2014-10-05 21:22 (UTC)
    - Fix tests tickling deficient threads on perl 5.10.0

0.10011 2014-09-26 11:24 (UTC)
    - Soft-depend on newer (bugfixed and *simpler*) Class::XSAccessor 1.19
    - More robust threading tests

0.10010 2013-04-24 02:58 (UTC)
    - Fix bug with identically-named 'simple' accessors in different
      classes set to access *differently named fields* getting their
      field access mixed up
    - Fix subtle pessimization when having identically-named accessors
      in different classes leads to 'simple' implementations not being
      replaced by Class::XSAccessor where appropriate

0.10009 2012-11-15 18:51 (UTC)
    - Stop leaking extra methods into the inheritance chain - there
      are plenty already

0.10008 2012-11-15 09:48 (UTC)
    - Allow disabling of accessor name checking introduced in 0.10007
    - Pass tests if Class::XSAccessor is available but Sub::Name isn't

0.10007 2012-11-08 11:54 (UTC)
    - Disable tests on perls where Class::XSAccessor emits broken
      error messages (RT#74883, RT#80519)
    - Drop minimum perl to 5.6 (from 5.6.2)
    - Switch all module loading to Module::Runtime and lose
      dependency on Class::Inspector
    - Fix stupid mistake causing double-require of Sub::Name when
      Class::XSAccessor is not available (RT#80657)
    - Simplify superclass traversal done by the 'inherited' group type
    - Fix incorrect quoting of unusual hash keys (fieldnames)
    - Depend on newer bugfixed Class::XSAccessor 1.15
    - Improve text of ro/wo violation exceptions
    - Sanity-check accessor names for well-formedness
      (qr/[A-Z_a-z][0-9A-Z_a-z]*/)

0.10006 2011-12-30 03:52 (UTC)
    - Silence warnings resulting from incomplete can() overrides
      hiding get/set_simple methods

0.10005 2011-12-26 12:43 (UTC)
    - Depend on newer bugfixed Class::XSAccessor
    - Repack with correct metadata (RT#73100)

0.10004 2011-11-28 21:20 (UTC)
    - No longer leak internal __CAG* methods into the inheritable
      namespace

0.10003 2011-05-03 00:15 (UTC)
    - Only require MRO::Compat for older perls
    - Add SYNOPSIS
    - Add examples for methods that get used most often

0.10002 Sun Dec 19 05:23:44 2010
    - Fix grave bug of XS-enabled simple accessors clobbering an
      existing 'around' overlay installed in the same method slot
    - Require bugfixed XSAccessor, remove Win32 caveat

0.10001 Sun Dec 12 03:17:05 2010
    - Fix an ActiveState Win32 incompatibility
    - Fix spurious method re-invocation warnings after Class::Unload

0.10000 Sat Nov 27 17:51:04 2010
    - Fix perl 5.6 failures
    - Add test-time deferred coderef reinvocation checks
    - Another minor (8%) speedup

0.09009 Fri Nov 26 01:31:56 2010
    - Major cleanup and optimization of code (evaled coderef sharing)
    - Module can now operate in PurePerl environments with 100%
      compatibility (including proper naming of generated coderefs)

0.09008 Sun Oct 11 07:41:56 2010
    - Put back a private undocumented method that the DBIC-CDBI compat
      layer relies on :(
    - Fix corner case segfaults with C::XSA and old 5.8 perls

0.09007 Sat Oct  9 10:22:56 2010 (DELETED)
    - Fix corner case when get/set_simple overrides are circumvented
      iff Class::XSAccessor is present

0.09006 Fri Sep 10 23:55:00 2010
    - Fix bugs in ro/wo accessor generation when XSAccessor is
      being used
    - Better Class::XSAccessor usage control - introducing
      $ENV{CAG_USE_XS} and $Class::Accessor::Grouped::USE_XS

0.09005 Wed Sep  1 04:00:00 2010
    - Again, remove Class::XSAccessor for Win32 sine it still breaks

0.09004 Wed Aug 11 04:23:15 2010
    - Changed the way Class::XSAccessor is invoked if available
      (recommended by C::XSA author)
    - Modified internal cache names to avoid real accessor clashes
    - Some micro-optimizations for get_inherited
    - Fixed field names with a single quote in them (patch from Jason Plum)

0.09003 Fri Apr 23 23:00:19 2010
    - use Class::XSAccessor if available for 'simple' accessors, except on
      MSWin32, with documentation

0.09002 Tue Oct 20 23:16:28 2009
    - removing Class::XSAccessor usage for now

0.09001 Thu Oct  1 21:48:06 2009
    - remove optional dep from Makefile.PL

0.09000 Sun Aug 23 20:08:09 2009
    - release

0.08999_01 Tue July 7 22:06:21 2009
    - Make _mk_group_accessors name the closures installed for Moose compat
    - Use Class::XSAccessor if available RT#45577 (Andy Grundman)

0.08003 Sat Mar 21 9:27:24 2009
    - Fixed set_inherited under C3::Componentised: RT#43702, RIBASUSHI

0.08002 Mon Nov 17 20:27:22 2008
    - Removed unnecessary code in get_simple: RT#40992, BUCHMULLER Norbert

0.08001 Wed Jan 09 19:35:34 2008
    - Fixed Makefile.PL tests setting that was killing older installs

0.08000 Tue Jan 08 18:22:47 2008
    - Bumped version for release. No changes oherwise.

0.07009_01 Fri Dec 28 18:08::00 2007
    - Tweak code for pure speed while fixing performance issue when assigning @_
    under Perl 5.10.0

0.07000
    - Altered get_inherited to return undef rather than () when no value
        set for Class::Data::(Inheritable|Accessor) compatiblity
    - Fixed spelling test error
    - Added WriteAll/DIST/PREOP for README

0.06000 Fri May 11 22:00:26 2007
    - get_super_paths now uses mro::get_linear_isa to DTRT under C3

0.05002 Fri May 11 20:46:16 2007
    - killed Class::Inspector->installed warnings

0.05001 Thur May 10 20:55:11 2007
    - set_component_class now only dies if the specified class is a
        installed/installable class and fails to load it.

0.05000 Tue May 08 19:42:33 2007
    - Added get/set_component_class


t/lib/AccessorGroups.pm
{
  package AccessorGroups::BeenThereDoneThat;

  use strict;
  use warnings;
  use base 'Class::Accessor::Grouped';

  __PACKAGE__->mk_group_accessors('simple', 'singlefield');
  __PACKAGE__->mk_group_accessors('multiple', qw/multiple1 multiple2/);
}


package AccessorGroups;
use strict;
use warnings;
use base 'Class::Accessor::Grouped';
__PACKAGE__->mk_group_accessors('simple', 'singlefield');
__PACKAGE__->mk_group_accessors('multiple', qw/multiple1 multiple2/);
__PACKAGE__->mk_group_accessors('listref', [qw/lr1name lr1;field/], [qw/lr2name lr2'field/]);
__PACKAGE__->mk_group_accessors('simple', 'runtime_around');
__PACKAGE__->mk_group_accessors('simple', [ fieldname_torture => join ('', map { chr($_) } (0..255) ) ]);

sub get_simple {
  my $v = shift->SUPER::get_simple (@_);
  $v =~ s/ Extra tackled on$// if $v;
  $v;
}

sub set_simple {
  my ($self, $f, $v) = @_;
  $v .= ' Extra tackled on' if $f eq 'singlefield';
  $self->SUPER::set_simple ($f, $v);
  $_[2];
}

# a runtime Class::Method::Modifiers style around
# the eval/our combo is so that we do not need to rely on Sub::Name being available
my $orig_ra_cref = __PACKAGE__->can('runtime_around');
our $around_cref = sub {
  my $self = shift;
  if (@_) {
    my $val = shift;
    $self->$orig_ra_cref($val . ' Extra tackled on');
    $val;
  }
  else {
    my $val = $self->$orig_ra_cref;
    $val =~ s/ Extra tackled on$// if defined $val;
    $val;
  }
};
{
  no warnings qw/redefine/;
  eval <<'EOE';
    sub AccessorGroups::runtime_around { goto $AccessorGroups::around_cref };
    sub AccessorGroups::_runtime_around_accessor { goto $AccessorGroups::around_cref };
EOE
}

sub new {
  return bless {}, shift;
};

foreach (qw/multiple listref/) {
  no strict 'refs';
  *{"get_$_"} = __PACKAGE__->can('get_simple');
  *{"set_$_"} = __PACKAGE__->can('set_simple');
};

1;


t/lib/AccessorGroupsRO.pm
package AccessorGroupsRO;
use strict;
use warnings;
use base 'Class::Accessor::Grouped';

__PACKAGE__->mk_group_ro_accessors('simple', 'singlefield');
__PACKAGE__->mk_group_ro_accessors('multiple', qw/multiple1 multiple2/);
__PACKAGE__->mk_group_ro_accessors('listref', [qw/lr1name lr1;field/], [qw/lr2name lr2'field/]);
__PACKAGE__->mk_group_ro_accessors('simple', [ fieldname_torture => join ('', map { chr($_) } (1..255) ) ]);

sub new {
  return bless {}, shift;
};

foreach (qw/multiple listref/) {
  no strict 'refs';
  *{"get_$_"} = __PACKAGE__->can ('get_simple');
};

1;


t/lib/AccessorGroupsWO.pm
package AccessorGroupsWO;
use strict;
use warnings;
use base 'Class::Accessor::Grouped';

__PACKAGE__->mk_group_wo_accessors('simple', 'singlefield');
__PACKAGE__->mk_group_wo_accessors('multiple', qw/multiple1 multiple2/);
__PACKAGE__->mk_group_wo_accessors('listref', [qw/lr1name lr1;field/], [qw/lr2name lr2'field/]);
__PACKAGE__->mk_group_wo_accessors('simple', [ fieldname_torture => join ('', map { chr($_) } (1..255) ) ]);

sub new {
  return bless {}, shift;
};

foreach (qw/multiple listref/) {
  no strict 'refs';
  *{"set_$_"} = __PACKAGE__->can('set_simple');
};

1;


Makefile.PL
use strict;
use warnings;
use 5.006;
use lib 'inc';
use ExtUtils::HasCompiler qw(can_compile_loadable_object);
use ExtUtils::MakeMaker;

my %META = (
  name => 'Class-Accessor-Grouped',
  license => 'perl_5',
  author => [ 'Matt S. Trout <mst@shadowcatsystems.co.uk>' ],
  prereqs => {
    configure => { requires => {
      'ExtUtils::MakeMaker'   => 0,
    } },
    runtime => {
      requires => {
        'perl'            => '5.006',
        'Carp'            => 0,
        'Module::Runtime' => '0.012',
        'Scalar::Util'    => 0,
      },
      recommends => {
        'Sub::Name'         => '0.05',
        # when changing CXSA version don't forget to adjust lib/Class/Accessor/Grouped.pm as well
        'Class::XSAccessor' => '1.19',
      },
    },
    test => {
      requires => {
        'Test::More' => '0.88',
        'Test::Exception' => '0.31',
      },
    },
    develop => {
      requires => {
        'Sub::Name'           => '0.05',
        'Class::XSAccessor'   => '1.19',
        'Pod::Coverage'       => '0.14',
        'Test::CheckManifest' => '0.09',
        'Test::NoTabs'        => '0.03',
        'Test::Pod'           => '1.00',
        'Test::Pod::Coverage' => '1.04',
        'Test::Spelling'      => '0.11',
        'Test::Strict'        => '0.05',
      },
    },
  },
  resources => {
    homepage => 'https://metacpan.org/release/Class-Accessor-Grouped',
    repository => {
      url => 'git://git.shadowcat.co.uk/p5sagit/Class-Accessor-Grouped.git',
      web => 'https://github.com/p5sagit/Class-Accessor-Grouped',
      type => 'git',
    },
    bugtracker => {
      web => 'https://rt.cpan.org/Public/Dist/Display.html?Name=Class-Accessor-Grouped',
      mailto => 'bug-Class-Accessor-Grouped@rt.cpan.org',
    },
    license => [ 'http://dev.perl.org/licenses/' ],
  },
  no_index => {
    directory => [ 't', 'xt' ]
  },
);

my %MM_ARGS = (
  PREREQ_PM => {
    (
      (
        parse_args()->{PUREPERL_ONLY} 
        or !can_compile_loadable_object(quiet => 1)
      ) ? ()
      : (
        'Sub::Name' => $META{prereqs}{runtime}{recommends}{'Sub::Name'},
        # CXSA does not work on 5.6
        ( "$]" > 5.008 ? (
          'Class::XSAccessor' => $META{prereqs}{runtime}{recommends}{'Class::XSAccessor'}
        ) : ()),
      )
    ),
    ("$]" < 5.009_005 ? ( 'MRO::Compat' => 0 ) : ()),
  },
  realclean => { FILES => 'README' },
  (-e 'META.yml' ? () : test => { TESTS => 't/*.t xt/*.t xt/*/*.t' } ),
);

{
  package MY;
  sub distdir {
    my ($self, @arg) = @_;
    my $frag = $self->SUPER::distdir(@arg);
    $frag =~ s/^(distdir\s*:.*)/$1 create_readme/m;
    $frag .= <<'END_MAKE';
create_readme: create_distdir
	pod2text lib/Class/Accessor/Grouped.pm > $(DISTVNAME)/README
	cd $(DISTVNAME) && $(PERLRUN) "-MExtUtils::Manifest=maniadd" -e "maniadd({ README => 'Generated README' })"
END_MAKE
    $frag;
  }
}

use Text::ParseWords;

sub parse_args {
  # copied from EUMM
  ExtUtils::MakeMaker::parse_args(
    my $tmp = {},
    Text::ParseWords::shellwords($ENV{PERL_MM_OPT} || ''),
    @ARGV,
  );
  return $tmp->{ARGS} || {};
}

## BOILERPLATE ###############################################################
require ExtUtils::MakeMaker;

# have to do this since old EUMM dev releases miss the eval $VERSION line
my $eumm_version  = eval $ExtUtils::MakeMaker::VERSION;
my $mymeta        = $eumm_version >= 6.57_02;
my $mymeta_broken = $mymeta && $eumm_version < 6.57_07;

($MM_ARGS{NAME} = $META{name}) =~ s/-/::/g;
($MM_ARGS{VERSION_FROM} = "lib/$MM_ARGS{NAME}.pm") =~ s{::}{/}g;
$MM_ARGS{ABSTRACT_FROM} = $MM_ARGS{VERSION_FROM};
$META{license} = [ $META{license} ]
  if $META{license} && !ref $META{license};
$MM_ARGS{LICENSE} = $META{license}[0]
  if $META{license} && $eumm_version >= 6.30;
$MM_ARGS{NO_MYMETA} = 1
  if $mymeta_broken;
$MM_ARGS{META_ADD} = { 'meta-spec' => { version => 2 }, %META }
  unless -f 'META.yml';

for (qw(configure build test runtime)) {
  my $key = $_ eq 'runtime' ? 'PREREQ_PM' : uc $_.'_REQUIRES';
  my $r = $MM_ARGS{$key} = {
    %{$META{prereqs}{$_}{requires} || {}},
    %{delete $MM_ARGS{$key} || {}},
  };
  defined $r->{$_} or delete $r->{$_} for keys %$r;
}

$MM_ARGS{MIN_PERL_VERSION} = delete $MM_ARGS{PREREQ_PM}{perl} || 0;

delete $MM_ARGS{MIN_PERL_VERSION}
  if $eumm_version < 6.47_01;
$MM_ARGS{BUILD_REQUIRES} = {%{$MM_ARGS{BUILD_REQUIRES}}, %{delete $MM_ARGS{TEST_REQUIRES}}}
  if $eumm_version < 6.63_03;
$MM_ARGS{PREREQ_PM} = {%{$MM_ARGS{PREREQ_PM}}, %{delete $MM_ARGS{BUILD_REQUIRES}}}
  if $eumm_version < 6.55_01;
delete $MM_ARGS{CONFIGURE_REQUIRES}
  if $eumm_version < 6.51_03;

ExtUtils::MakeMaker::WriteMakefile(%MM_ARGS);
## END BOILERPLATE ###########################################################
